<launch>

<param name="use_sim_time" value="true" />

<!--
   <node
     name="dynamixel_manager"
     pkg="dynamixel_controllers"
     type="controller_manager.py"
     output="screen">
 
     <rosparam>
       namespace: dxl_manager
       serial_ports:
         pan_tilt_port:
           port_name: "/dev/ttyUSB0"
           baud_rate: 1000000
           min_motor_id: 1
           max_motor_id: 2
           update_rate: 60
     </rosparam>

   </node>

   <rosparam
      file="$(find spin_hokuyo)/launch/dynamixel_servos_tilting.yaml"
      command="load"
    />

  <node
    name="dynamixel_controller_spawner"
    pkg="dynamixel_controllers"
    type="controller_spawner.py"
    args="manager=dxl_manager
          port=pan_tilt_port
          type=simple
          tilt_controller"
    output="screen"
   />

  <node
    name="tilt_motor"    
    pkg="spin_hokuyo"
    type="tilt_motor"
    output="screen"
   />

  <node
    name="tilt_transform"
    pkg="spin_hokuyo"
    type="tilt_transform"
    output="screen"
  />

 <node
    pkg="hokuyo_node"
    type="hokuyo_node"
    name="hokuyo"
    output="screen">

    <rosparam>
      min_ang: -2.0071
      max_ang: 2.0071
    </rosparam>
  </node>

-->

<node
   pkg="spin_hokuyo"
   type="hokuyo_robot_filter"
   name="hokuyo_robot_filter"
   output="screen"
/>



  <node
    pkg="rviz"
    type="rviz"
    name="rviz"
    args="-d $(find motion_hokuyo)/rviz_cfg/removed_floor.rviz"/>
   />

  <node pkg="tf" type="static_transform_publisher" name="servo_to_velodyne"
            args="0.35 0 -0.13 0 0 -0.052 /velodyne /servo 10" />


  <node
    pkg="spin_hokuyo"
    type="scan_to_pcl"
    name="scan_to_pcl"
   />


  
    <node 
        name="relay_hokuyo" pkg="topic_tools" type="relay"
        args="/hokuyo_points /hokuyo_velodyne_combined_cloud"
    />

    <node 
        name="relay_velodyne" pkg="topic_tools" type="relay"
        args="/velodyne_points /hokuyo_velodyne_combined_cloud"
    />

    <param name="assembled_cloud_mode" type="string" value="time" />
    <param name="scan_time" type="double" value="0.3" />
    <arg name="assemble" default="velodyne_points"/> 

  <node type="point_cloud2_assembler" pkg="laser_assembler" name="point_cloud_assembler">
    <remap from="cloud" to="$(arg assemble)"/>
    <param name="max_clouds" type="int" value="10000" />
    <param name="fixed_frame" type="string" value="map" />
  </node>


   <node
     pkg="spin_hokuyo"
     type="pcl_assembler_client"
     name="pcl_assembler_client"
     output="screen"
    />
 
    <node pkg="octomap_server" type="octomap_server_node" name="octomap_server">
         <param name="resolution" value="0.3" />
         <param name="frame_id" type="string" value="map" />
         <param name="sensor_model/max_range" value="10.0" />
         <remap from="cloud_in" to="/pcl_test" />
    </node>

    <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="screen" />

    <node pkg="nodelet" type="nodelet" name="voxel_grid" args="load pcl/VoxelGrid pcl_manager" output="screen">
      <remap from="~input" to="/assembled_cloud" />
      <rosparam>
        filter_field_name: z
        filter_limit_min: -10000
        filter_limit_max: 10000
        filter_limit_negative: False
        leaf_size: 1
      </rosparam>
    </node>

    <node pkg="motion_hokuyo" type="hokuyo_ground_filter" name="hokuyo_ground_filter"/>

    <node pkg="motion_hokuyo" type="plane_extract_ransac" name="plane_extract_ransac">
    </node>

    <node pkg="motion_hokuyo" type="max_z" name="max_z">
    	<remap from="assembled_cloud" to="/planes_pointcloud"/>
    </node>

    <arg name="path"/> 
    <arg name="speed" default="1"/> 
    <arg name="other" default=""/> 
    <arg name="topics" default=""/> 
    <node pkg="rosbag" type="play" name="player" output="screen" args="$(arg path) -r $(arg speed) --topics tf velodyne_points scan --clock" />

</launch>
